core/src/main/java/org/opensearch/sql/datasource/DataSourceServiceImpl.java

```diff
@@ -140,17 +173,37 @@ private Optional<DataSourceMetadata> getDataSourceMetadata(String dataSourceName
 
   private DataSource getDataSourceFromMetadata(DataSourceMetadata dataSourceMetadata) {
     if (!dataSourceMap.containsKey(dataSourceMetadata)) {
-      clearDataSource(dataSourceMetadata);
+      clearDataSourceWithSameName(dataSourceMetadata.getName());
```

Comment:
why we need to remove first if the given data source doesn't exist?

Reply:
dataSourceMap is used to cache Datasource Objects against datasource Metadata object.

So let's say I have a datasource with name my_prometheus.
* In the first call, the datasource object[not metadata] is cached against the datasource metadata object
* If in the mean time someone updates metadata in another node.
* In the next call for the datasource, it fetches the metadata again from index and figures out there is no cache available for new metadata.
* So I cache the newly created datasource object and caches it against newly created metadata object. I will also cleanup the old cache entry.

let me know if this is confusing and there is a better way to do this.

Reply:
Seems we need DataSourceCache?
```
class DataSourceCache {
   DataSource getDataSource(DataDourceMetaData metadata);
   void putDataSource(DataDourceMetaData metadata)
   void invalidateDataSource(DataDourceMetaData metadata);
}

// DataSourceMeataData should include version info.
DataDourceMetaData {
  int hashCode() {
     return version + context;
  }
}

// also include DataSource CacheBuilder to configure the cache behaviour
```

Reply:
Having Lombok annotation on DataSourceMetadata will include all the fields in hashcode generation. Therefore, two different objects with same data will have same hashcode and it fits our requirement.  Let me know if I am missing anything here.
My requirement is simple with hashmap features. Will create a cache builder if requirement increases to complex expiration strategies. I will be leveraging Google Guava Cache so I can expand easily on it.

For now, I am Creating a separate class for Cache with our specific requirements.